include/merlin/core_kernels/lookup_kernels.cuh

```diff
@@ -0,0 +1,722 @@
+
+// Using 32 threads to deal with one key
+template <typename K = uint64_t, typename V = float, typename S = uint64_t,
+          typename CopyScore = CopyScoreEmpty<S, K, 128>,
+          typename CopyValue = CopyValueTwoGroup<float, float4, 32>>
+__global__ void lookup_kernel_with_io_pipeline_v1(
+    const Bucket<K, V, S>* buckets, const size_t buckets_num, const int dim,
+    const K* __restrict keys, V* __restrict values, S* __restrict scores,
+    bool* __restrict founds, size_t n) {
+  constexpr int GROUP_SIZE = 32;
+  constexpr int RESERVE = 16;
+  constexpr int DIM_BUF = 224;
+  constexpr int BLOCK_SIZE = 128;
+  constexpr int BUCKET_SIZE = 128;
+  constexpr int GROUP_NUM = BLOCK_SIZE / GROUP_SIZE;
+  constexpr int DIGEST_SPAN = BUCKET_SIZE / 4;
+
+  __shared__ int sm_target_digests[BLOCK_SIZE];
+  __shared__ K sm_target_keys[BLOCK_SIZE];
+  __shared__ K* sm_keys_ptr[BLOCK_SIZE];
+  __shared__ V* sm_values_ptr[BLOCK_SIZE];
+  // Reuse
+  S* sm_target_scores = reinterpret_cast<S*>(sm_target_keys);
+  int* sm_counts = sm_target_digests;
+  int* sm_founds = sm_counts;
+  // Double buffer
+  __shared__ uint32_t sm_probing_digests[2][GROUP_NUM * DIGEST_SPAN];
+  __shared__ K sm_possible_keys[2][GROUP_NUM * RESERVE];
+  __shared__ int sm_possible_pos[2][GROUP_NUM * RESERVE];
+  __shared__ V sm_vector[2][GROUP_NUM][DIM_BUF];
+
+  // Initialization
+  auto g = cg::tiled_partition<GROUP_SIZE>(cg::this_thread_block());
+  int groupID = threadIdx.x / GROUP_SIZE;
+  int rank = g.thread_rank();
+  int key_idx_base = (blockIdx.x * blockDim.x) + groupID * GROUP_SIZE;
+  if (key_idx_base >= n) return;
+  int loop_num =
+      (n - key_idx_base) < GROUP_SIZE ? (n - key_idx_base) : GROUP_SIZE;
+  if (rank < loop_num) {
+    int idx_block = groupID * GROUP_SIZE + rank;
+    K target_key = keys[key_idx_base + rank];
+    sm_target_keys[idx_block] = target_key;
+    K hashed_key = Murmur3HashDevice(target_key);
+    uint8_t target_digest = static_cast<uint8_t>(hashed_key >> 32);
+    sm_target_digests[idx_block] = static_cast<uint32_t>(target_digest);
+    int global_idx = hashed_key % (buckets_num * BUCKET_SIZE);
+    int bkt_idx = global_idx / BUCKET_SIZE;
+    const Bucket<K, V, S>* bucket = buckets + bkt_idx;
+    __pipeline_memcpy_async(sm_keys_ptr + idx_block, &(bucket->keys_),
+                            sizeof(K*));
+    __pipeline_commit();
+    __pipeline_memcpy_async(sm_values_ptr + idx_block, &(bucket->vectors),
+                            sizeof(V*));
+  }
+  __pipeline_wait_prior(0);
+
+  // Pipeline loading
+  S* scores_ptr =
+      reinterpret_cast<S*>(sm_keys_ptr[groupID * GROUP_SIZE] + BUCKET_SIZE);
+  uint8_t* digests_ptr = reinterpret_cast<uint8_t*>(scores_ptr + BUCKET_SIZE);
+  __pipeline_memcpy_async(sm_probing_digests[0] + groupID * DIGEST_SPAN + rank,
+                          digests_ptr + rank * 4, sizeof(uint32_t));
+  __pipeline_commit();
+  __pipeline_commit();  // padding
+  __pipeline_commit();  // padding
```

Comment:
@jiashuy Why do you need these paddings?

Reply:
I think the `__pipeline_wait_prior(x)` waits for the (x+1)th `__pipeline_commit()` in front.
You can observe the `__pipeline_wait_prior(3)` at line 109.
So, in the first loop, we need to wait `sm_probing_digests` to be writen back at  the stage of pipeline loading.
So I pad `__pipeline_commit()` to avoid to check in the loop again and again.